Backups Dashboard: add restore feature

Initsogar opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the “fixes” keyword. Consider using “part of” instead.
-->

Part of DOTDASH-278

## Proposed Changes

* Add Backup's restore feature to the Hosting Dashboard
  * Add `siteBackupRestoreInitiateMutation` to handle restore initiation
  * Add `siteBackupRestoreProgressQuery` to handle restore progress
  * Add DataForm for selecting the restore configuration
  * Add form, in progress, success and failure steps

> [!NOTE]
> There may be some adjustments to make on the success and failure state design. I will follow up on this in a separate PR.

### Demo

https://github.com/user-attachments/assets/0229413a-ca5f-4573-93f5-2c7ec819683a

### Screenshots

| Screen | Capture |
|---|---|
| Form / Initial state | <img width="500" alt="CleanShot 2025-08-20 at 22 48 30@2x" src="https://github.com/user-attachments/assets/bdb30c76-8793-40f2-884a-d730434a0d8f" /> |
| In progress state | <img width="500" alt="CleanShot 2025-08-20 at 22 49 19@2x" src="https://github.com/user-attachments/assets/d0ede649-5787-4727-b52e-349b40a092df" /> |
| Success state | <img width="500" alt="CleanShot 2025-08-20 at 22 50 20@2x" src="https://github.com/user-attachments/assets/c7813e81-d461-40a0-ae3e-7510bd536687" /> |
| Failure state | <img width="500"  alt="CleanShot 2025-08-20 at 22 49 58@2x" src="https://github.com/user-attachments/assets/72c4e023-f0a3-4bfe-8dd4-7c2bfa45b5d4" /> |

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Part of Backups in the Hosting Dashboard

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
“Before / After" screenshots can also be very helpful when the change is visual.
-->

* Navigate to `/v2/sites/SITE_SLUG/backups` using a site with the backup feature
* Click on any of the backups on the list, then click on -> Restore to this point
* It should redirect you to /v2/sites/ragostini.com/backups/#####/restore
* Now you are in the initial state (the form)
* Select the items you would like to restore
  * If you deselect “Site dabase” the notice at the bottom changes slightly :)
* When you are ready, click on the `Restore now` button.
* Now you are in the in-progress state.
* Wait for the restore to complete.
* After completion, you should see the `Site successfully restored` message with an option to see all backups (go back to the main Backup page) or an external link to visit your site.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the “[Status] String Freeze” label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the “[Status] Needs Privacy Updates” label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Initsogar committed:
Add data layer functions for restore feature

* Implement siteBackupRestoreProgressQuery to fetch restore progress.
* Create siteBackupRestoreInitiateMutation to initiate a site restore.
* Define RestoreConfig and RestoreProgress types in data layer.
* Add API functions for initiating restores and fetching progress.

Initsogar committed:
Add Site Backup Restore Form Component

* Implement SiteBackupRestoreForm for initiating site restores.
* Include form fields for selecting restore options (themes, plugins, etc.).
* Add validation and user notifications for restore actions.
* Integrate with existing site backup restore API functions.

Initsogar committed:
Add Site Backup Restore Progress Component

* Implement SiteBackupRestoreProgress to display the status of site restore operations.
* Fetch and manage restore progress data.
* Include visual indicators such as progress percentage and messages.

Initsogar committed:
Add Site Backup Restore Success Component

* Implement SiteBackupRestoreSuccess to display a success message after site restoration.
* Include buttons for viewing the website and accessing all backups.

Initsogar committed:
Add Site Backup Restore Error Component

* Implement SiteBackupRestoreSuccess to display an error message when site restoration fails.
* Include a retry button for users to attempt the restore again or contact support.

Initsogar committed:
Implement Site Backup Restore Component

arthur791004 commented on client/dashboard/app/queries/site-backup-restore.ts:
```diff
@@ -0,0 +1,47 @@
+		queryKey: [ 'site', siteId, 'backup', 'restore', restoreId, 'progress' ],
+		queryFn: () => fetchSiteBackupRestoreProgress( siteId, restoreId ),
+		enabled: !! restoreId,
+		refetchInterval: ( query: { state: { data?: RestoreProgress } } ) => {
```
I'm curious about whether the type is required or not. I guess `refetchInterval` knows the type automatically 🤔

Initsogar replied:
I was circling on this a bit. It looks like the type isn’t inferred automatically because refetchInterval’s callback is typed against a generic `Query` object that isn’t connected to our `queryFn` return type. That means `query.state.data` defaults to `unknown` unless we add an explicit type.

By narrowing the parameter to `{ state: { data?: RestoreProgress } }`, we get type safety on just the piece we care about (`data.status`) without having to deal with the full `Query` type.

arthur791004 commented on client/dashboard/sites/backup-restore/progress.tsx:
```diff
@@ -0,0 +1,70 @@
+			<Spacer marginTop={ 12 }>
+				<Notice variant="info" title={ __( 'Check your email' ) }>
+					{ __(
+						"Don't want to wait? For your convenience, we'll email you when your site has been fully restored."
```
We should use curly quotes `‘`

Initsogar replied:
Done in 94d6a36d3ab6bc0879ed2442c82c43f6d45ac448

arthur791004 commented on client/dashboard/sites/backup-restore/error.tsx:
```diff
@@ -0,0 +1,23 @@
+import { __ } from '@wordpress/i18n';
+import Notice from '../../components/notice';
+
+const SiteBackupRestoreSuccess = ( { onRetry }: { onRetry: () => void } ) => {
```
We should change the name 🙂

Initsogar replied:
Yep, totally missed it :)
Thank you, addressed it on 0796acbd6bfed977b9fde527978c222345347e93

arthur791004 commented on client/dashboard/sites/backup-restore/form.tsx:
```diff
@@ -0,0 +1,149 @@
+	const handleFormChange = ( changes: Partial< RestoreConfig > ) => {
+		setFormData( ( prev ) => ( {
+			...prev,
+			...Object.fromEntries(
```
Is there any reason we need to check whether the value is `undefined` or not 🤔 Instead of

```js
onChange={ ( edits: Partial<RestoreConfig> ) => {
	setFormData( ( data ) => ( { ...data, ...edits } ) );
}
```

Initsogar replied:
Not at all, just simplified it on b88683d4e2feed1f02f99d72b34a2f11f1763aaa

arthur791004 commented on client/dashboard/sites/backup-restore/progress.tsx:
```diff
@@ -0,0 +1,70 @@
+					{ isRunning ? restoreProgress?.message : __( 'Initializing the restore process' ) }
+				</Text>
+				<Text size={ 13 } variant="muted">
+					{ restoreProgress?.percent }% completed
```
We should translate it 🙂

Initsogar replied:
oops! Addressed it on ddbebc1e31160386d77bb373064af41635cfe229

taipeicoder commented on client/dashboard/app/queries/site-backup-restore.ts:
```diff
@@ -0,0 +1,47 @@
+	queryOptions( {
+		queryKey: [ 'site', siteId, 'backup', 'restore', restoreId, 'progress' ],
+		queryFn: () => fetchSiteBackupRestoreProgress( siteId, restoreId ),
+		enabled: !! restoreId,
```
`enabled` seems to be already being passed down in https://github.com/Automattic/wp-calypso/pull/105301/files#diff-ac80dd13a8c40b4703012a39b007e8e0fc4d76afff96e4beb450c16dcd8d6420R28. WDYT having it only at the component-layer?

Initsogar replied:
Nice catch! Sounds good to me.
Addressed in 695069c895c0bd4cb441ee18587f713d20aaae19.

Initsogar committed:
Remove redundant type annotation from refetchInterval callback

Initsogar committed:
Use curly quotes in user-facing text

Initsogar committed:
Fix component name to reflect error state

Initsogar committed:
Simplify form change handler logic

Initsogar committed:
Add translation support for completion percentage text

Initsogar committed:
Remove duplicate enabled prop from query definition

Keep enabled control at component level only to avoid redundant query configuration.

Initsogar committed:
Revert "Remove redundant type annotation from refetchInterval callback"

This reverts commit 7ef0ad331915f18aa4bb8d2629e0f766e01ba6a8.

myhro reviewed (approved):
This is looking really good. Thanks for implementing that, Rafa. LGTM. ✅
